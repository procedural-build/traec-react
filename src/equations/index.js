import React from "react";
import { BSBtnDropdown } from "traec-react/utils/bootstrap";
import { getAllOps } from "traec/utils/metricOperations";
import Octicon from "react-octicon";

export const renderOps = (ops, baseMetrics) => {
  for (let op of ops) {
    let metricIds = op.metricList ? op.metricList.match(/[0-9a-f-]{36}/g) || [] : [];
    let renderedMetrics = metricIds
      .map(id => baseMetrics.get(id))
      .filter(i => i != null)
      .map((bm, i) => (
        <li key={i}>
          {bm ? (
            <span>
              {bm.get("name")} {/*[{bm.get("uid").substring(0, 8)}]*/}
            </span>
          ) : null}
        </li>
      ));
    Object.assign(op, {
      metricIds,
      renderedMetrics
    });
  }
  return ops;
};

export const getMetricsFromIndicator = ({ indicator, baseMetrics }) => {
  let text = indicator.getInPath("operation.text");
  let ops = getAllOps(text);
  // Get the overall factor
  let factor = ops[0].factor;
  // Append the list of metricComponents to the operations
  ops = renderOps(ops, baseMetrics);
  return { factor, ops };
};

export function TargetValueDisplay({ targetValue, raThreshold, greenBelow }) {
  if (!targetValue) {
    return null;
  }
  let arrowUp = raThreshold ? raThreshold < targetValue : !greenBelow;
  let arrow = arrowUp ? <Octicon name="arrow-up" /> : <Octicon name="arrow-down" />;
  return (
    <React.Fragment>
      {arrow} {targetValue} {raThreshold ? `| ${raThreshold}` : null}
    </React.Fragment>
  );
}

export function OperationMetrics({ ops, numerator = true }) {
  let lists = [];
  let validOps = ops.filter(i => i.exponent > 0 == numerator && i.renderedMetrics.length);
  let multipleSets = validOps.length > 1;
  let counter = 0;
  for (let op of validOps) {
    lists.push(
      <React.Fragment key={counter}>
        {multipleSets ? (
          <p className="m-0 p-0">
            <b>{Math.sign(op.factor) > 0 ? "Add" : "Subtract"}</b>
          </p>
        ) : null}
        <ul style={{ margin: 0, padding: 0 }}>{op.renderedMetrics}</ul>
      </React.Fragment>
    );
    counter += 1;
  }
  if (!numerator && lists.length == 0) {
    lists = "1";
  }
  return <React.Fragment>{lists}</React.Fragment>;
}

export function IndicatorEqn({ ops }) {
  return (
    <React.Fragment>
      <ul style={{ margin: 0, padding: 0 }}>
        <OperationMetrics ops={ops} />
      </ul>
      <hr style={{ margin: "0.25em", padding: 0, border: null, borderBottom: "1px solid black" }} />
      <ul style={{ margin: 0, padding: 0 }}>
        <OperationMetrics ops={ops} numerator={false} />
      </ul>
    </React.Fragment>
  );
}

export default function IndicatorRow({ baseMetrics, indicator, dropDownLinks, setTargetComponent }) {
  if (!baseMetrics) {
    return null;
  }
  let { factor, ops } = getMetricsFromIndicator({ indicator, baseMetrics });

  let metricTarget = indicator.get("metricTarget");
  let targetValue = metricTarget ? metricTarget.get("value") : null;
  let raThreshold = metricTarget ? metricTarget.getInPath("meta_json.thresholdLow") : null;
  let greenBelow = metricTarget ? metricTarget.getInPath("meta_json.greenBelow") : null;

  // Useful for injecting into the UI for debugging
  let idStr = (
    <span>
      [{indicator.get("uid").substring(0, 8)}:{indicator.getInPath("resultBaseMetric.uid").substring(0, 8)}]
    </span>
  );

  return (
    <React.Fragment>
      <div className="row">
        <div className="col-sm-4">
          <span className="align-middle">{indicator.getInPath("resultBaseMetric.name")}</span>
        </div>
        <div className="col-sm-2 align-middle">{indicator.getInPath("resultBaseMetric.category")}</div>
        <div className="col-sm-1 align-middle">{factor ? parseFloat(factor).toFixed(2) : factor}</div>
        <div className="col-sm-3 align-middle">
          <IndicatorEqn ops={ops} />
        </div>
        <div className="col-sm-1 align-middle">
          <TargetValueDisplay targetValue={targetValue} raThreshold={raThreshold} greenBelow={greenBelow} />
        </div>
        <div className="col-sm-1 align-middle">
          <BSBtnDropdown
            links={dropDownLinks}
            header={
              <span>
                <Octicon name="gear" />
              </span>
            }
          />
        </div>
      </div>
      {/*this.render_add_form(indicator, factor, numIds, denIds)*/}
      {setTargetComponent}
      <hr />
    </React.Fragment>
  );
}
