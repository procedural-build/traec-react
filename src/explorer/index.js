import React from "react";
import { connect } from "react-redux";
import Traec from "traec";
import CategoryRow from "./category";

class TrackerTree extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      formParams: {
        initFields: {}
      }
    };

    // Don't include edges in this call because response takes too long
    this.requiredFetches = [
      new Traec.Fetch("tracker", "read")
      /*new Traec.Fetch("tracker_branch", "list", {}, {
        preUpdateHook: (args) => ({...args, include_edges: true})
      })*/
    ];

    this.addRootCategory = this.addRootCategory.bind(this);
  }

  componentDidMount() {
    Traec.fetchRequired.bind(this)();
  }

  componentDidUpdate() {
    Traec.fetchRequired.bind(this)();
  }

  dropDownLinks() {
    return [{ name: "Add a Category", onClick: this.addRootCategory }];
  }

  addRootCategory(e) {
    e.preventDefault();
    let { trackerId, refId, commitId, treeId } = this.props;
    let fetch = new Traec.Fetch("tracker_ref_tree_branch", "post", {
      trackerId,
      refId,
      commitId,
      treeId
    });
    this.setState({ formParams: { params: fetch.params } });
    fetch.toggleForm();
  }

  render() {
    const {
      tracker,
      trackerId,
      refId,
      commitId,
      renderRootTree = true,
      showTreesWithoutDescriptions = true,
      formFields = null
    } = this.props;

    if (!tracker) {
      return null;
    }

    return (
      <React.Fragment>
        <CategoryRow
          isRoot={true}
          extraRowClass={"ml-0"}
          tracker={tracker}
          trackerId={trackerId}
          commitId={commitId}
          branchId={null}
          refId={refId}
          renderRootTree={renderRootTree}
          rootTreeName={tracker.get("name")}
          showTreesWithoutDescriptions={showTreesWithoutDescriptions}
          forceExpand={true}
          addWithDescriptions={false}
          formFields={formFields}
        />
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  const { trackerId } = ownProps;
  let tracker = state.getInPath(`entities.trackers.byId.${trackerId}`);
  let refId = tracker ? tracker.getInPath("root_master") : null;
  let ref = state.getInPath(`entities.refs.byId.${refId}`);
  let commitId = ref ? ref.getInPath(`latest_commit.uid`) : null;
  let treeId = ref ? ref.getInPath(`latest_commit.tree_root.uid`) : null;
  return { trackerId, tracker, refId, ref, commitId, treeId };
};

const mapDispatchToProps = dispatch => ({ dispatch });

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TrackerTree);
