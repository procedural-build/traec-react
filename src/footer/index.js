import React from "react";

export class Footer extends React.Component {
  render() {
    return (
      <div className="footer">
        <div className="footer-inner">
          <p>News</p>
          <p>About</p>
          <p>Help</p>
        </div>
        <div className="footer-tools">
          <span className="go-top" />
        </div>
      </div>
    );
  }
}
