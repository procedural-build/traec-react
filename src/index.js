import * as utils from './utils';
import * as auth from './auth';


export default {
    utils,
    auth,
}

exports.printMsg = function() {
    console.log("This is a message from the demo package");
  }