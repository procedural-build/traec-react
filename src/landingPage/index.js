import React from "react";
import ReactMarkdown from "react-markdown";

export class TraecLandingPage extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    return <StaticContent content={this.props.content} />;
  }
}

export const StaticContent = function(content = null) {
  if (!content) {
    content =
      "ODS Track makes gathering, review, follow-up, emailing, analysis, compilation and submission of documents for certification easy.";
  }
  return (
    <React.Fragment>
      <ReactMarkdown source={content.content} />
    </React.Fragment>
  );
};
