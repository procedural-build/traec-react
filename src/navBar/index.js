import React from "react";
import { connect } from "react-redux";
import Im from "traec/immutable";
import DropdownLogin from "./loginDropdown";
import { MenuItem } from "./item";
import { DropDownItem } from "./dropdown";
import { setNavBarItems } from "traec/redux/actionCreators";

export { setNavBarItems };

export class NavBar extends React.Component {
  renderItems() {
    if (this.props.items == null) {
      return null;
    }
    return this.props.items.map((item, i) => renderItem(item, i));
  }

  renderBrand() {
    let { brand } = this.props;
    if (brand) {
      return brand;
    }
    return (
      <React.Fragment>
        Traec | <b>Procedural</b>.build
      </React.Fragment>
    );
  }

  render() {
    let { extraClass } = this.props;
    return (
      <div className="navbar-area">
        <nav
          className={`navbar navbar-expand-lg navbar-light bg-light navbar-static-top ${extraClass}`}
          role="navigation"
        >
          <div className="container-fluid m-0 p-0">
            <div className="navbar-brand">{this.renderBrand()}</div>

            <button
              className="navbar-toggler border-0"
              type="button"
              data-toggle="collapse"
              data-target="#exCollapsingNavbar"
            >
              &#9776;
            </button>

            <div className="collapse navbar-collapse" id="exCollapsingNavbar">
              <ul className="nav navbar-nav flex-row justify-content-between ml-auto">
                {this.renderItems()}
                {this.props.preUserItems}
                <DropdownLogin />
              </ul>
            </div>
          </div>
        </nav>
      </div>
    );
  }
}

const renderItem = function(item, keyIndex) {
  const label = item.get("label");
  const to = item.get("to");

  if (Im.List.isList(to)) {
    return <DropDownItem key={keyIndex} label={label} items={to} />;
  } else {
    return <MenuItem key={keyIndex} label={label} to={to} item={item} />;
  }
};

const mapStateToProps = (state, ownProps) => {
  return {
    items: state.getInPath(`ui.navbar.items`)
  };
};

const mapDispatchToProps = dispatch => {
  return {
    dispatch: dispatch
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(NavBar);
