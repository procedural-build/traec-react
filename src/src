/Users/MLadmin/Applications/actionSustainability/npm/traec-react/src