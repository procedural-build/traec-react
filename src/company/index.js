import React from "react";
import ReactDOM from "react-dom";
import Traec from "traec";

import Sidebar from "AppSrc/sidebar";
import { Footer } from "AppSrc/footer";
import { connect } from "react-redux";

import { setNavBarItems } from "traec/redux/actionCreators";
import { setSidebarItems } from "AppSrc/sidebar/actionCreators";

import {
  companyPermissionRender,
  getCompanyPermissions,
  companyPermissionFilter,
  companyPermissionCheck
} from "traec/utils/permissions/company";
import { CompanyTree } from "./tree";
// import { CompanyRouter } from "./router";

import BootstrapSplitPane from "../utils/bootstrap/splitbs";

/**
 * Company Page:
 * @namespace CompanyPage
 * @example
 * * <Route path="/company/:companyId" component={Company} />
 */
class CompanyPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      setNavBar: false,
      setSideBar: false,
      showSideBar: true
    };

    // Data required from the API for this Component
    this.requiredFetches = [new Traec.Fetch("company", "list")];

    // Action bindings
    this.navbarLinks = this.navbarLinks.bind(this);
    this.sidebarLinks = this.sidebarLinks.bind(this);
  }

  componentDidMount() {
    let { companyId } = this.props;
    companyPermissionCheck(companyId, false, []);
    Traec.fetchRequired.bind(this)();
    this.setNavBar();
  }

  componentDidUpdate(prevProps) {
    let { companyId } = this.props;
    companyPermissionCheck(companyId, false, []);
    Traec.fetchRequired.bind(this)();

    if (companyId && companyId != prevProps.companyId) {
      this.setState({
        setSideBar: false,
        setNavBar: false
      });
    }

    this.setSideBar();
    this.setNavBar();
  }

  setNavBar(forceUpdate = false) {
    let navBarLinks = this.navbarLinks();
    if ((!this.state.setNavBar && navBarLinks) || forceUpdate) {
      this.setState({ setNavBar: true });
      this.props.dispatch(setNavbarItems(navBarLinks));
    }
  }

  setSideBar(forceUpdate = false) {
    let sideBarLinks = this.sidebarLinks();
    if ((!this.state.setSideBar && sideBarLinks) || forceUpdate) {
      this.setState({ setSideBar: true });
      this.props.dispatch(setSidebarItems(sideBarLinks));
    }
  }

  navbarLinks() {
    let { companyId } = this.props;
    let companyBase = `/company/${companyId}`;

    let items = [
      {
        label: "Company",
        requiresAdmin: false,
        to: [
          { label: "Company Dashboard", to: companyBase, octicon: "home" },
          {
            label: "Members",
            to: `${companyBase}/members`,
            octicon: "organization",
            requiresAdmin: false,
            requiredActions: ["READ_COMPANY_MEMBER"]
          }
        ]
      },
      {
        label: "Admin",
        requiresAdmin: true,
        to: [
          { label: "Indicators", to: `${companyBase}/indicators`, octicon: "issue-closed", requiresAdmin: true },
          { label: "Settings", to: `${companyBase}/details`, octicon: "gear", requiresAdmin: true },
          { label: null },
          { label: "Email Settings", to: `${companyBase}/email/settings`, octicon: "inbox", requiresAdmin: true },
          { label: "Email Statistics", to: `${companyBase}/email/report`, octicon: "mail", requiresAdmin: true }
        ]
      }
    ];
    return companyPermissionFilter(companyId, items);
  }

  render_main() {
    let { companyId, company, companyList } = this.props;
    if (!company) {
      return "";
    }

    return (
      <React.Fragment>
        <div className="container-fluid">
          <div className="row">
            <BootstrapSplitPane
              localStorageKey={`explorer-grid-split-${companyId}`}
              pane1ClassName={"page-sidebar"}
              onCollapseHook={() => {
                this.setState({ showSideBar: false });
              }}
              onExpandHook={() => {
                this.setState({ showSideBar: true });
              }}
              pane1Style={{
                borderRight: "1px solid grey",
                minHeight: "800px"
              }}
            >
              <div>{this.render_sidebar()}</div>
              {/* <div>
                <CompanyRouter companyId={companyId} company={company} />
              </div> */}
            </BootstrapSplitPane>
          </div>
        </div>
        <Footer />
      </React.Fragment>
    );
  }
  render_sidebar() {
    let { companyId, companyList, company } = this.props;
    let companyBase = `/company/${companyId}`;

    if (!this.state.showSideBar) {
      return null;
    }
    return (
      <React.Fragment>
        <Sidebar />
        <hr />
        <CompanyTree company={company} companyList={companyList} currentId={companyId} />
      </React.Fragment>
    );
  }
  sidebarLinks() {
    let { companyId } = this.props;
    let companyBase = `/company/${companyId}`;
    let items = [
      { label: "Company Dashboard", to: companyBase, octicon: "home" },
      {
        label: "Members",
        to: `${companyBase}/members`,
        octicon: "organization",
        requiresAdmin: false,
        requiredActions: ["READ_COMPANY_MEMBER"]
      },
      { label: "Indicators", to: `${companyBase}/indicators`, octicon: "issue-closed", requiresAdmin: true },
      { label: "Settings", to: `${companyBase}/details`, octicon: "gear", requiresAdmin: true }
    ];
    return companyPermissionFilter(companyId, items);
  }
  render() {
    // Check the User permissions for this company
    return companyPermissionRender(this.props.companyId, false, ["READ_COMPANY_REPORT"], this.render_main(), true);
  }
}

const mapStateToProps = (state, ownProps) => {
  let { companyId } = Traec.utils.getFullIds(state, ownProps.match.params);

  let company = state.getInPath(`entities.companies.byId.${companyId}`);
  let companyList = state.getInPath("entities.companies.byId");
  let userPermissions = getCompanyPermissions(state, companyId);

  return { companyId, company, companyList, userPermissions };
};

const mapDispatchToProps = dispatch => {
  return {
    dispatch: dispatch
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(CompanyPage);
