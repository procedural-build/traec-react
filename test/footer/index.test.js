import React from "react";
import { shallow } from "enzyme";
import { TraecFooter } from "../../src/footer";
import TestRenderer from "react-test-renderer";
jest.mock("react-router-dom", () => ({ Link: "Link" }));

describe("Footer", () => {
  it("should render three <Link />s", () => {
    const wrapper = shallow(<TraecFooter />);
    expect(wrapper.find("Link").length).toEqual(3);
    expect(wrapper.find("Link").get(0).props.to).toEqual("/news");
    expect(wrapper.find("Link").get(1).props.to).toEqual("/about");
    expect(wrapper.find("Link").get(2).props.to).toEqual("/help");
  });

  it("should match snapshot", () => {
    const testRenderer = TestRenderer.create(<TraecFooter />);
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
