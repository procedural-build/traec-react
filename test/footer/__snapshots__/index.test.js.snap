// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer should match snapshot 1`] = `
<div
  className="footer fixed-bottom"
>
  <div
    className="footer-inner row"
  >
    <div
      className="col-sm-3"
    />
    <Link
      className="col-sm-3"
      to="/news"
    >
      News
    </Link>
    <Link
      className="col-sm-3"
      to="/about"
    >
      About
    </Link>
    <Link
      className="col-sm-3"
      to="/help"
    >
      Help
    </Link>
  </div>
  <div
    className="footer-tools"
  >
    <span
      className="go-top"
    />
  </div>
</div>
`;
